src/widget/wlibrarytableview.cpp

```diff
@@ -92,42 +65,112 @@ void WLibraryTableView::moveSelection(int delta) {
         return;
     }
 
-    while(delta != 0) {
-        // TODO(rryan) what happens if there is nothing selected?
-        QModelIndex current = currentIndex();
-        if(delta > 0) {
-            // i is positive, so we want to move the highlight down
-            int row = current.row();
-            if (row + 1 < pModel->rowCount()) {
-                selectRow(row + 1);
-            }
+    while (delta != 0) {
+        QItemSelectionModel* currentSelection = selectionModel();
+        if (currentSelection->selectedRows().length() > 0) {
+            if (delta > 0) {
+                // i is positive, so we want to move the highlight down
+                int row = currentSelection->selectedRows().last().row();
+                if (row + 1 < pModel->rowCount()) {
+                    selectRow(row + 1);
+                } else {
+                    // we wrap around at the end of the list so it is faster to get
+                    // to the top of the list again
+                    selectRow(0);
+                }
+
+                delta--;
+            } else {
+                // i is negative, so move down
+                int row = currentSelection->selectedRows().first().row();
+                if (row - 1 >= 0) {
+                    selectRow(row - 1);
+                } else {
+                    selectRow(pModel->rowCount() - 1);
+                }
 
-            delta--;
+                delta++;
+            }
         } else {
-            // i is negative, so we want to move the highlight up
-            int row = current.row();
-            if (row - 1 >= 0) {
-                selectRow(row - 1);
+            // no selection, so select the first or last element depending on delta
+            if (delta > 0) {
+                selectRow(0);
+                delta--;
+            } else {
+                selectRow(pModel->rowCount() - 1);
+                delta++;
             }
-
-            delta++;
         }
     }
 }
 
-void WLibraryTableView::saveVScrollBarPos(TrackModel* key){
-    m_vScrollBarPosValues[key] = verticalScrollBar()->value();
+void WLibraryTableView::saveTrackModelState(
+        const QAbstractItemModel* model, const QString& key) {
+    //qDebug() << "saveTrackModelState:" << model << key;
+    VERIFY_OR_DEBUG_ASSERT(model) {
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(!key.isEmpty()) {
+        return;
+    }
+    ModelState* state = m_modelStateCache.take(key);
```

Comment:
well, the original commit 49d263eb6587fd1ca13e63e375c5df5977feb0d0 is tagged "Make QCache usage safe against concurrent deletion", though I don't see when that would happen.
I guess @poelzi can explain his motivation for this change.

Reply:
This one is open. I think we should use the normal pattern for QCache here unless we can remember the reason and put that into the code.

Reply:
It follow the logic that the Object in the QCache is only ever owned by one owner, rust style. It is either in the update handler or in the QCache object, but it is never in the update handler while the QCache thinks it needs to drop this object.

Reply:
okay. could you add a review/code suggestion that adds a comment with this explanation?

Reply:
The Warning applies applies, if you store the pointer in a member variable, or if we use the cache from multiple threads. 
The QCrash has no magic to do something at laterally any time. It can do something at any call. 
All this does not apply here. So I am not able to write a reasonable comment. I assume the crashes was cased by something else.  

However since this is only a minor issue, and not on a performance critic path, I think we should keep this tested variant, but not add a comment with probably wrong assumptions.